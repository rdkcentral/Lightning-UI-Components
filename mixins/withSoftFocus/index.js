import { PALETTE } from '../../Styles';

export default function withSoftFocus(Base) {
  return class extends Base {
    static _template() {
      return {
        flex: {
          alignItems: 'center',
          direction: 'row',
          justifyContent: 'space-between'
        },
        SoftFocusMarker: {
          zIndex: 2,
          h: 45,
          w: 4,
          alpha: 0,
          rect: true,
          color: PALETTE.white
        },
        ...super._template()
      };
    }

    _init() {
      super._init();
      this._softFocused = false;
    }

    _focus() {
      super._focus();
      this._softFocused = true;
      this._SoftFocusMarker.patch({ alpha: 0 });
    }

    _unfocus() {
      super._unfocus();
      this._softFocused
        ? this._SoftFocusMarker.patch({ alpha: 1 })
        : this._SoftFocusMarker.patch({ alpha: 0 });
    }

    set softFocused(softFocused) {
      this._softFocused = softFocused;
      this._softFocused && !this.hasFocus()
        ? this._SoftFocusMarker.patch({ alpha: 1 })
        : this._SoftFocusMarker.patch({ alpha: 0 });
    }

    get softFocused() {
      return this._softFocused;
    }

    get _SoftFocusMarker() {
      return this.tag('SoftFocusMarker');
    }
  };
}
