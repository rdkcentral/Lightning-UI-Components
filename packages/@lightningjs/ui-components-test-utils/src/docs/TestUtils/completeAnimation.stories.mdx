<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / completeAnimation" />

# completeAnimation

Returns a Promise that resolves once all animating properties have updated to their target value(s).

## Arguments

`completeAnimation(element, transitionProperties = [])`

| argument             | type              | default | description                                                 |
| -------------------- | ----------------- | ------- | ----------------------------------------------------------- |
| element              | Lightning.Element | -       | element with properties that transition value changes       |
| transitionProperties | array\|string     | []      | property name(s) that transition value changes on component |

<br />

##### Example: Testing element properties before and after their values have transitioned

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    return {
      TransitioningRect: {
        rect: true,
        color: 0xffffffff,
        w: 100,
        h: 100,
        x: 0,
        y: 0
      }
    };
  }

  _updatePosition() {
    this.tag('TransitioningRect').setSmooth('x', this.axisPosition);
    this.tag('TransitioningRect').setSmooth('y', this.axisPosition);
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this._updatePosition();
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. await the transitions of x and y values to finish before testing the result
it('should transition the x value of the component', async () => {
  const [example] = createExampleComponent();
  expect(example.tag('TransitioningRect').x).toBe(0);
  expect(example.tag('TransitioningRect').y).toBe(0);

  example.axisPosition = 50;
  await completeAnimation(example.tag('TransitioningRect'), ['x', 'y']);

  expect(example.tag('TransitioningRect').x).toBe(50);
  expect(example.tag('TransitioningRect').y).toBe(50);
});
```
