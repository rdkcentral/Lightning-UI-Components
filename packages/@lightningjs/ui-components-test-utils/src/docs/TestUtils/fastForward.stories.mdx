<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / fastForward" />

# fastForward

Force all running Lightning Transitions on one or more Lightning Elements to finish and update the transitioning property to its target value immediately.

This function is similar to [completeAnimation](/story/unit-testing-test-utils-completeanimation--page) in that it allows testing transitioning properties. Unlike `completeAnimation`, `fastForward` is not asynchronous, and does result in all running transitions to complete. If a test requires awaiting specific properties to finish transitioning, `completeAnimation` is the recommended utility function. The `fastForward` function forces transitions to finish, but does not wait for the template to update before proceeding through the test function. To ensure the template has updated following a call to `fastForward`, invoke [`testRenderer.update()`](/story/unit-testing-test-renderer-overview--page).
// TODO: add link directly to testRenderer.update() docs once those are written

## Arguments

`fastForward(elements)`

| argument | type                                           | default | description                                   |
| -------- | ---------------------------------------------- | ------- | --------------------------------------------- |
| elements | Array<Lightning.Element\> \| Lightning.Element | -       | element(s) with transitioning property values |

<br />

##### Example: Forcing running Transitions to finish before testing the result

```js
import lng from '@lightningjs/core';
import {
  makeCreateComponent,
  fastForward
} from '@lightningjs/ui-components-test-utils';

class Example extends lng.Component {
  static _template() {
    const square = {
      rect: true,
      w: 100,
      h: 100,
      x: 0,
      y: 0,
      color: 0xeeeeeeee
    };
    return {
      TransitioningRectA: square,
      TransitioningRectB: square
    };
  }

  set axisPosition(value) {
    this._axisPosition = value;
    this.tag('TransitioningRectA').setSmooth('x', value);
    this.tag('TransitioningRectB').setSmooth('y', value);
  }

  get axisPosition() {
    return this._axisPosition;
  }
}

// generate a createComponent function to use in tests
const createExampleComponent = makeCreateComponent(Example);

// Ex. force all running transitions to complete, then test the result
it('should transition the x value of the component', () => {
  const [example, testRenderer] = createExampleComponent();
  expect(example.tag('TransitioningRectA').x).toBe(0);
  expect(example.tag('TransitioningRectB').y).toBe(0);

  //
  example.axisPosition = 50;
  const rectA = example.tag('TransitioningRectA');
  const rectB = example.tag('TransitioningRectB');
  fastForward([rectA, rectB]);
  testRenderer.update();

  expect(example.tag('TransitioningRectA').x).toBe(50);
  expect(example.tag('TransitioningRectB').y).toBe(50);
});
```
