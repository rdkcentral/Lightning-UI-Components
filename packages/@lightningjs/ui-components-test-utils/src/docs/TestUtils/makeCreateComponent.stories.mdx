<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Docs / Unit Testing / Test Utils / makeCreateComponent " />

# makeCreateComponent

Returns a function, referred to as a `createComponent` function, which is used to generate an instance of a component to run unit test cases against.

## Arguments

`makeCreateComponent(component, defaultConfig, defaultOptions)`

| argument       | type          | default                                  | description                                                                                                                |
| -------------- | ------------- | ---------------------------------------- | -------------------------------------------------------------------------------------------------------------------------- |
| component      | lng.Component | the component to generate an instance of | component to render an instance of                                                                                         |
| defaultConfig  | object        | {}                                       | properties that should be applied to the component by default                                                              |
| defaultOptions | object        | {}                                       | default options used when creating the component instance (these options are passed to the `TestRenderer.create` function) |

### `createComponent`

Returns an array of 2 elements:

1. an instance of the component to be tested against (generated by `testRenderer.getInstance()`)
2. a [testRenderer](/story/unit-testing-test-renderer-create--page#testrenderer-functions) object generated using the `component` argument and a merged object created from `defaultOptions` and `options`

#### Arguments

| argument | type   | default | description                                                                                                                                                                |
| -------- | ------ | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| config   | object | {}      | properties that should be applied to the component. This object is spread over the `defaultConfig` object.                                                                 |
| options  | object | {}      | options used when creating the component instance (these options are passed to the `TestRenderer.create` function). This object is spread over the `defaultOptions` object |

<br />

##### Example: Generating a `makeComponent` function and test component

```js
import lng from '@lightningjs/core';
import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';

// Lightning component to write unit tests for
class Example extends lng.Component {
  static _template() {
    return {
      Example: {
        text: { text: '' }
      }
    };
  }

  set exampleProp(value) {
    this._exampleProp = value;
    this.tag('Example').text.text = value;
  }

  get exampleProp() {
    return this._exampleProp;
  }
}

// unit tests for Example component

// generate a createComponent function to use in tests
const defaultConfig = { exampleProp: 'foo' };
const defaultOptions = { focused: false };
const createExampleComponent = makeCreateComponent(
  Example,
  defaultConfig,
  defaultOptions
);

// Ex. An instance of the component rendered with the properties defined in the defaultConfig and defaultOptions object
it('should render the exampleProp value as text', () => {
  const [example] = createExampleComponent();

  expect(example.exampleProp).toBe('foo');
  expect(example.tag('Example').text.text).toBe('foo');
  expect(example.hasFocus()).toBe(false);
});

// Ex. An instance of the component rendered with the properties defined in the config object
it('should support updating the text', () => {
  // overwrite the value of exampleProp defined in defaultConfig
  const [example] = createExampleComponent({
    exampleProp: 'bar'
  });

  expect(example.exampleProp).toBe('bar');
  expect(example.tag('Example').text.text).toBe('bar');
});

// Ex. An instance of the component rendered with options defined in the options object
it('should support rendering the component in a focused state', () => {
  // overwrite the value of focused defined in defaultOptions
  const [example] = createExampleComponent({}, { focused: true });

  expect(example.hasFocus()).toBe(true);
});

// Ex. Using a function provided by the testRenderer object (focus)
it('should allow focusing on the component', () => {
  const [example, testRenderer] = createExampleComponent();

  // focused was set to false in defaultOptions, so the componet will initially render without focus
  expect(example.hasFocus()).toBe(false);

  testRenderer.focus();

  expect(example.hasFocus()).toBe(true);
});
```

### spyOnMethods

A utility which enables running component methods asynchronously.

The `options` argument passed to the `createComponent` function accepts a field called `spyOnMethods`.
The value for `spyOnMethods` can be an array of method names to generate promises from. When one of those methods is invoked in the component it's associated promise will resolve, then the actual method will be invoked. The promises generated from the passed in methods are given the name `_${methodName}SpyPromise`.

<br />

##### Example: Spying on component methods and awaiting their execution before making assertions on their side effects

```js
class MethodSpiesExample extends Base {
  static get __componentName() {
    return 'MethodSpiesExample';
  }

  static _template() {
    return {
      Title: {
        type: TextBox
      },
      Subtitle: {
        type: TextBox,
        y: 100
      }
    };
  }

  static get tags() {
    return ['Title', 'Subtitle'];
  }

  static get properties() {
    return ['title', 'subtitle'];
  }

  _construct() {
    super._construct();
    this._title = 'title';
    this._subtitle = 'subtitle';
  }

  _update() {
    this._updateTitle();
    this._updateSubtitle();
  }

  _updateTitle() {
    this._Title.content = this.title;
  }
  _updateSubtitle() {
    this._Subtitle.content = this.subtitle;
  }
}

const createComponent = makeCreateComponent(MethodSpies);

it('should update the title and subtitle', async () => {
  const [methodSpiesComponent, testRenderer] = createComponent(
    {},
    { spyOnMethods: ['_update'] }
  );

  // wait for _update to be called on the initial render before proceeding
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe('title');
  expect(methodSpiesComponent._Subtitle.content).toBe('subtitle');

  const newTitle = 'new title';
  const newSubtitle = 'new subtitle';
  methodSpiesComponent.patch({
    title: newTitle,
    subtitle: newSubtitle
  });

  // wait for _update to be called in response to values in the properties array changing
  await methodSpiesComponent.__updateSpyPromise;

  expect(methodSpiesComponent._Title.content).toBe(newTitle);
  expect(methodSpiesComponent._Subtitle.content).toBe(newSubtitle);
});
```

**Note: `spyOnMethods` can only be set in in the `options` object of `createComponent`. It can not be assigned to the `defaultOptions` object in `makeCreateComponent`**
