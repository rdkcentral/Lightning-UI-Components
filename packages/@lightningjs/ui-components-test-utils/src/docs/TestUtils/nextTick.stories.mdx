<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

import { Meta } from '@storybook/addon-docs';

<Meta title="Unit Testing / Test Utils / nextTick" />

# nextTick

Creates a Promise that resolves after a defined amount of time. If no amount of time is specified, the Promise will resolve immediately.

## Arguments

`nextTick(wait = 0)`

| argument | type   | default | description                                                           |
| -------- | ------ | ------- | --------------------------------------------------------------------- |
| wait     | number | 0       | amount of time (in milliseconds) before the returned Promise resolves |
