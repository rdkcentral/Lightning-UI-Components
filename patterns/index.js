export { default as BoardRow } from './BoardRow';
export {
  default as Card,
  BasicCard,
  HorizontalCard,
  HorizontalCardLarge,
  VerticalCard,
  VerticalCardLarge,
  VerticalCardDynamic
} from './Card';
export { default as DataItem } from './DataItem';
export { default as GameStats } from './GameStats';
export { default as Info } from './Info';
export { default as Input } from './Input';
export { default as Keyboard, KEYBOARD_FORMATS } from './Keyboard';
export {
  default as ListItem,
  ListItemImage,
  ListItemImageGroup,
  ListItemSlider,
  ListItemToggle,
  ListItemRadio,
  ListItemPicker
} from './ListItem';
export { default as Notification } from './Notification';
export { default as OverlayDataItem } from './OverlayDataItem';
export { default as TeamInfo } from './TeamInfo';
export { default as TeamMatchup } from './TeamMatchup';
export { default as withBadgeProgress } from './withBadgeProgress';
export { default as withMetadata } from './withMetadata';
