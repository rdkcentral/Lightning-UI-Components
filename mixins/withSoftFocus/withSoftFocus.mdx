import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import withSoftFocus from '.';

# withSoftFocus

A `withSoftFocus` mixin, which decorates [`ListItem`](?path=/docs/listitem--basic) with specific marker,
indicating that even when ui focus is lost, marked [`ListItem`](?path=/docs/listitem--basic) is still in *selected* state.

## Source

https://github.comcast.com/Lightning/lightning-ui/packages/@lightning/ui-core/src/components/withSoftFocus/withSoftFocus.js

## Usage

`withSoftFocus` marker is being represented by configurable white dash, attached to the left side of `ListItem` component.

```js
import { withSoftFocus } from '@lightning/ui';
import { ListItem } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      ListItem: {
          type: withSoftFocus(ListItem),
          backgroundType: args.backgroundType,
          title: args.title,
          subtitle: args.subtitle,
          size: args.size,
          softFocused: args.softFocus
        }
    };
  }
};
```

Previews:

<Canvas>
  <Story id="mixins-withsoftfocus--basic" />
</Canvas>

Soft Focus Marker will appear only in case ListItem component does not have actual ui focus. There is no need to display

additional markers in case ui focus has already covered current ListItem .

## API

### Properties

| name           | type                                         | readonly | description
| -------------- | -------------------------------------------- | -------- | ----------------------------------------------------------- |
|softFocused      |boolean                                      | false    | defines if current ListItem should be marked with SoftFocus


### Methods

#### _init(): void

Called each time on component initialization. Manages initial value of `softFocused` and calls super._init()

#### _focus(): void

Called each time component gains focus. Manages value of `softFocused` and visibility of softFocus marker. Calls super._focus()

#### _unfocus(): void

Called each time component loses focus. Manages value of `softFocused` and visibility of softFocus marker. Calls super._focus()

### Styles

Styled components can be adjusted by composing using [withStyles](?path=/docs/mixins-withstyles--basic). Below is the default `styles` object

<Theme theme={withSoftFocus.styles} />
