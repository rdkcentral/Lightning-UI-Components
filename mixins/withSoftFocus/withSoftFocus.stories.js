import lng from '@lightningjs/core';

import { ListItem } from '../../patterns';
import mdx from './withSoftFocus.mdx';
import withSoftFocus from './index';

export default {
  title: 'mixins/withSoftFocus',
  args: {
    title: 'List Item',
    subtitle: 'List Item Metadata',
    size: 'small',
    backgroundType: 'fill',
    softFocus: true
  },
  argTypes: {
    backgroundType: {
      control: {
        type: 'radio',
        options: ['fill', 'float', 'ghost']
      }
    },
    focused: { control: 'boolean' },
    title: { control: 'text' },
    subtitle: { control: 'text' },
    size: {
      control: {
        type: 'radio',
        options: ['xsmall', 'small', 'medium', 'large']
      }
    },
    softFocus: { control: 'boolean' }
  },
  parameters: {
    tag: 'ListItem',
    argActions: {
      softFocus: (softFocus, component) => {
        const listItem = component.tag('ListItem');
        listItem.softFocused = softFocus;
      },
      focused: (isFocused, component) => {
        component._getFocused = isFocused
          ? () => component.tag('ListItem')
          : () => {};
        component._refocus();
      }
    },
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ListItem: {
          type: withSoftFocus(ListItem),
          backgroundType: args.backgroundType,
          title: args.title,
          subtitle: args.subtitle,
          size: args.size,
          softFocused: args.softFocus
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('ListItem');
      }
    }
  };
