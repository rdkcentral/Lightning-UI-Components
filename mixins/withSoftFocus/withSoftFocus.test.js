import TestUtils from '../../test/lightning-test-utils';
import withSoftFocus from '.';
import { ListItem } from '../../patterns';

const createComponent = TestUtils.makeCreateComponent(
  withSoftFocus(ListItem),
  {},
  { focused: false }
);

describe('withSoftFocus', () => {
  let withSoftFocusListItem, testRenderer;

  beforeEach(() => {
    [withSoftFocusListItem, testRenderer] = createComponent();
  });
  afterEach(() => {
    withSoftFocusListItem = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });
  it('Initial state: not focused, not softFocused', () => {
    expect(withSoftFocusListItem.hasFocus()).toBe(false);
    expect(withSoftFocusListItem._softFocused).toBeDefined();
    expect(withSoftFocusListItem._softFocused).toBe(false);
    expect(withSoftFocusListItem._SoftFocusMarker).toBeDefined();
    expect(withSoftFocusListItem._SoftFocusMarker.alpha).toBe(0);
  });
  it('gain focus', () => {
    withSoftFocusListItem._focus();
    testRenderer.focus();
    expect(withSoftFocusListItem.hasFocus()).toBe(true);
    expect(withSoftFocusListItem._softFocused).toBeDefined();
    expect(withSoftFocusListItem._softFocused).toBe(true);
    expect(withSoftFocusListItem._SoftFocusMarker).toBeDefined();
    expect(withSoftFocusListItem._SoftFocusMarker.alpha).toBe(0);
  });
  it('lose focus', () => {
    withSoftFocusListItem._focus();
    testRenderer.focus();
    withSoftFocusListItem._unfocus();
    testRenderer.unfocus();
    expect(withSoftFocusListItem.hasFocus()).toBe(false);
    expect(withSoftFocusListItem._softFocused).toBeDefined();
    expect(withSoftFocusListItem._softFocused).toBe(true);
    expect(withSoftFocusListItem._SoftFocusMarker).toBeDefined();
    expect(withSoftFocusListItem._SoftFocusMarker.alpha).toBe(1);
  });
  it('lose softFocus', () => {
    withSoftFocusListItem._focus();
    testRenderer.focus();
    withSoftFocusListItem._unfocus();
    testRenderer.unfocus();
    withSoftFocusListItem.softFocused = false;
    expect(withSoftFocusListItem.hasFocus()).toBe(false);
    expect(withSoftFocusListItem._softFocused).toBeDefined();
    expect(withSoftFocusListItem._softFocused).toBe(false);
    expect(withSoftFocusListItem._SoftFocusMarker).toBeDefined();
    expect(withSoftFocusListItem._SoftFocusMarker.alpha).toBe(0);
  });
});
