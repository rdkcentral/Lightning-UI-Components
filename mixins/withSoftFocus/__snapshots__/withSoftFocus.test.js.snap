// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSoftFocus renders 1`] = `
Object {
  "Component": Object {
    "Container": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Left": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Left",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Left",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 2,
        },
        "Right": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": true,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Right",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Right",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 2,
        },
      },
      "clipping": false,
      "color": "ff232328",
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 96,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Container",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "SoftFocusMarker": Object {
      "active": false,
      "alpha": 0,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": false,
      "flex": false,
      "flexItem": true,
      "h": 45,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "SoftFocusMarker",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "SoftFocusMarker",
      ],
      "texture": Object {
        "type": "RectangleTexture",
      },
      "visible": true,
      "w": 4,
      "x": 0,
      "y": 0,
      "zIndex": 2,
    },
  },
}
`;
